### PR TITLE
Refactor, improvement and PHP 8.1 only

### PR DESCRIPTION
**Pull request by guidobonuzzi**
This PR includes:

- The constructor is now private
- Static constructor `between`, `after`, `before`, `forEmptySequence`, `fromString`
- Creating a Rank at the beginning or at the end is now more efficient
- `rankValidator` with new Exceptions
- Tests improvement
- Psalm
- Coding standard
- Infection wind Mutation Score Indicator (MSI) of 99%

A new major release is needed, compatibility is broken because the new constructor is private.

### CODE REVIEW

**Comment by Ocramius** [comment]
@alexcrawford to be clear, this PR is a proposal for a new **major** release of this library.

Up to you if you are interested in it: it was done in good faith, to improve the API, stability and behaviour of the library.

I asked @guidobonuzzi to contribute it back to your original project, rather than having us maintain a private fork: hope it is well perceived 👍

File: .run/Infection on entire Project.run.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<component name="ProjectRunConfigurationManager">
```
Comment (Ocramius) [inline .run/Infection on entire Project.run.xml]: These `.run` files come from your IDE, and shouldn't be part of the PR

File: infection.json.dist [hunk 2]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (Ocramius) [inline infection.json.dist:1]: This doesn't run in CI, does it?

File: psalm.xml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0"?>
```
Comment (Ocramius) [inline psalm.xml:1]: This isn't verified in CI, isn't it?

File: phpcs.xml.dist [hunk 4]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0"?>
```
Comment (Ocramius) [inline phpcs.xml.dist:1]: This isn't verified in CI, isn't it?

**Review by Ocramius — APPROVED** [review]
LGTM: mostly needs small CI and artifact -related changes

File: .run/Infection on entire Project.run.xml (hunk 1 above)
Comment (guidobonuzzi) [reply .run/Infection on entire Project.run.xml]: Yes, they came from PhpStorm. I'm removing them!

File: infection.json.dist (hunk 2 above)
Comment (guidobonuzzi) [reply infection.json.dist:1]: I want to create a CI. This is the next step.

**Comment by alexcrawford** [comment]
Looks interesting - I will do some integration testing tomorrow.

Is it good practice to commit `composer.lock` in a project like this?

**Comment by guidobonuzzi** [comment]
> Looks interesting - I will do some integration testing tomorrow.
> 
> Is it good practice to commit `composer.lock` in a project like this?

I added `composer.lock` to `.gitignore` to prevent commit, but I don't have a sure answer, probably you are right...

@Ocramius, what is the right answer?

**Comment by alexcrawford** [comment]
Ignore my last, I see it's still listed in .gitignore. I should have had my morning coffee before looking at it!

**Comment by Ocramius** [comment]
FWIW, I endorse committing it, but only in combination with lowest+latest CI pipelines.

The pipeline I proposed @guidobonuzzi to include yesterday has that feature 👍

**Comment by Ocramius** [comment]
CI won't work here until merged (due to GitHub security model - that's normal).

Meanwhile, here are the CI run results on the fork: https://github.com/KanbanBOX/lexorank-php/actions/runs/3198936467

**Comment by guidobonuzzi** [comment]
> CI won't work here until merged (due to GitHub security model - that's normal).
> 
> Meanwhile, here are the CI run results on the fork: https://github.com/KanbanBOX/lexorank-php/actions/runs/3198936467

It's impressive how simple it is to implement! Thank you @Ocramius

**Comment by Ocramius** [comment]
Worked on it for a couple years @_@

File: composer.json [hunk 5]
```diff
@@ -10,7 +10,7 @@
     ],
     "type": "project",
     "require": {
-        "php": ">=8.1",
+        "php": ">=8.0",
```
Comment (Ocramius) [inline composer.json]: Let's please use `^8.0` - we don't know if PHP 9.0 will be compatible :D 

I also endorse `~8.0 || ~8.1`, as explained in https://github.com/laminas/laminas-diactoros/issues/117#issuecomment-1267306142

Comment (Ocramius) [reply composer.json]: @guidobonuzzi important to fix this previous to merge BTW

Comment (Ocramius) [reply composer.json]: Handled in 6a3df83

**Review by Ocramius — APPROVED** [review]

**Comment by Ocramius** [comment]
That was quick :O

**Comment by alexcrawford** [comment]
Merged and released as `2.0.0`

**Comment by Ocramius** [comment]
Thanks @guidobonuzzi @alexcrawford! :partying_face:

**Comment by guidobonuzzi** [comment]
Thanks!

**Comment by boesing** [comment]
FYI as I stumbled upon this: ~8.0 === ^8.0

**Comment by Ocramius** [comment]
Argh, indeed @_@
